### PR TITLE
Download sample

### PR DESCRIPTION
**Pull request by kathryn1995**

### CODE REVIEW

File: api/juno/api/views.py [hunk 1]
```diff
@@ -1,23 +1,59 @@
...
 
 
 def download_file(request, laneId):
-    the_file = os.path.abspath(
+    lane_dir = os.path.abspath(
         os.path.join(
-            os.path.join(getattr(settings, "BASE_DIR", None),os.pardir
-            ),
-            "mock-data/" + "31663_7#113" + ".tar.gz",
+            os.path.join(getattr(settings, "BASE_DIR", None), os.pardir),
```
Comment (garethpeat) [inline api/juno/api/views.py:13]: This doesn't change per request, so could abstract calculation out to a constant at the top, eg.:
```
DATA_DIR = os.path.join(os.path.join(getattr(settings, "BASE_DIR", None), os.pardir), "mock-data")
```

File: api/juno/api/views.py [hunk 2]
```diff
@@ -1,23 +1,59 @@
...
+            "mock-data/" + "31663_7#113",
```
Comment (garethpeat) [inline api/juno/api/views.py:14]: We now we're going to change this, but for anyone who wasn't privy to the conversation, you could add a comment above, such as `TODO: replace fixed mock file with actual file for lane`.

**Comment by kathryn1995** [comment]
Gareth's above comments above need to be addressed, which are specific to the view functionality. 
Tests for the download button functionality still need to be written. Some of these can be written using Cypress. May need to keep a small mock file to run the download and check it has worked in the tests or one could be created each time the tests are run

File: api/juno/api/views.py [hunk 3]
```diff
@@ -1,23 +1,59 @@
...
+        "31663_7#113/spades_assembly/contigs.fa",
+        "31663_7#113/spades_assembly/annotation",
+        "31663_7#113/spades_assembly/annotation/31663_7#113.gff",
+    ]:
+        return thing
+
+
+def make_tarfile(output_filename, source_dir):
+    if os.path.exists(source_dir):
+        with tarfile.open(output_filename, "w:gz") as tar:
```
[the remark this answers is not in the document]
Comment (kathryn1995) [reply api/juno/api/views.py]: Tar file is now stored in memory

File: api/juno/api/views.py [hunk 4]
```diff
@@ -1,23 +1,60 @@
...
 from django.conf import settings
 
+# TODO: replace fixed mock file with actual file for lane
+DATA_DIR = os.path.join(
+    os.path.join(getattr(settings, "BASE_DIR", None), os.pardir),
+    "mock-data/" + "31663_7#113",
+)
+
 
 def download_file(request, laneId):
```
[the remark this answers is not in the document]
Comment (kathryn1995) [reply api/juno/api/views.py]: renamed view function and put StreamingHttpResponse into this function along with raising the 404 error here. Make tar file function in the downloads script takes laneID and returns BytesIO object which is passed to the Streamer

File: api/juno/api/views.py [hunk 5]
```diff
@@ -1,23 +1,60 @@
...
+        "31663_7#113/spades_assembly/annotation",
+        "31663_7#113/spades_assembly/annotation/31663_7#113.gff",
+    ]:
+        return thing
+
+
+def make_tarfile(output_filename, source_dir):
+    if os.path.exists(source_dir):
+        with tarfile.open(output_filename, "w:gz") as tar:
+            tar.add(
```
[the remark this answers is not in the document]
Comment (kathryn1995) [reply api/juno/api/views.py]: Files and their corresponding names are stored in a dictionary making it easy to add to and change

**Comment by garethpeat** [comment]
@kathryn1995, this is awesome.

I have one very minor suggestion about the `test_download_sample` and `test_download_sample_file_not_found`, which currently call the view function directly, since you're using `unittest.TestCase`. If you switch to subclassing `django.test.TestCase` and the provided `self.client` (see [docs](https://docs.djangoproject.com/en/3.0/topics/testing/tools/#default-test-client) and [download example](https://stackoverflow.com/questions/8244220/django-unit-test-for-testing-a-file-download)), it might make the tests very slightly cleaner. This could be done separately later though, in a different PR if you like, since your tests are functional.

**Comment by kathryn1995** [comment]
Added e2e test that button exists when the database is filled and is empty when the database is empty. 
Had to add "chromeWebSecurity": false to cypress.json to get click function to work. This may need changed. Looking at options to test download as it accesses different link.

File: e2e/integration/download-button.spec.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      cy.on("window:alert", stub);
+      cy.get(`table#sampleTable`)
+        .contains("td button", "31663_7#113")
+        .click()
+        .then(() => {
+          expect(stub.getCall(0)).to.be.calledWith(
+            "http://localhost:8001/SampleDownload/31663_7%23113,31663_7#113.tar.gz"
+          );
+        });
+      cy.wait(API_WAIT_MS);
```
Comment (garethpeat) [inline e2e/integration/download-button.spec.js:57]: I think you could remove this second wait.

File: ui/src/components/DownloadButton.js [hunk 7]
```diff
@@ -1,15 +1,21 @@
...
-	return (
-	  <Button onClick={handler} startIcon={<CloudDownloadIcon/>} >{children}</Button>
-	);
+const DownloadButton = ({ url, filename, children }) => {
+  const handler = () => {
+    if (window.Cypress) {
+      alert(url.concat("," + filename));
+      return;
+    }
+    FileSaver.saveAs(url, filename);
```
Comment (garethpeat) [inline ui/src/components/DownloadButton.js:12]: It's not great that we have to change the actual code for the test, but leave it in for now, as there doesn't seem to be an alternative. If we write a lot more tests around downloading in the future, we may need to look into running a non-headless browser, which does seem to support testing downloads.

It'd be helpful if you could add an info/warning message here, referencing any useful links you found on this topic (eg. https://github.com/cypress-io/cypress/issues/949 seems to be the main discussion).

File: e2e/integration/download-button.spec.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    loadDatabaseProfile(DB_PROFILES.SMALL).then((db) => {
+      // non-empty samples table?
+      expect(db.sample.length).to.be.greaterThan(0);
+
+      // load page
+      cy.visit("/");
+      cy.wait(API_WAIT_MS);
+      //   cy.get(`table#sampleTable`).contains("td button", "31663_7#113").click();
+      const stub = cy.stub();
+      cy.on("window:alert", stub);
```
Comment (garethpeat) [inline e2e/integration/download-button.spec.js:48]: Maybe also note briefly here that the `DownloadButton.onClick` handler is using custom `alert()` behaviour purely for the end-to-end tests, hence the stubbing.

**Pull request merged**